### ISSUE TITLE
Need to support addition of limited number of milestones for a timeline

### ISSUE THREAD

**Issue by vikasrohit** [topcoder-platform/tc-project-service#271]
For upcoming feature of project plan where we are going to improve the project plan, we are going to provide facility to add milestones to a timeline from front end. However, we don't want them to be able to add unlimited number of milestones. We want them to be able to add only configurable number of milestones. This configuration should be done at product template level.

fyi @maxceem @RishiRajSahu

**Comment by vikasrohit** [comment]
Results of our (me and @RishiRajSahu ) latest brainstorming about this topic:

1. We should have ability to have `metadata` for `productTemplates`
2. Add new property/field, `maxMilestones`, to `timeline` model
3. Populate this new field `maxMilestones` while creating the timeline for a product. Copy this value from `productTemplates`'s `metadata`
4. Update `PATCH` endpoint of `milestone` to respect the `maxMilestones` value from the `timeline`

@maxceem Let us know if you have any query about this. We would like to have this ability in place before we get final designs for new project plan changes where we are going to support addition of custom milestones in phases.

**Comment by maxceem** [comment]
@vikasrohit I have a couple of questions regarding `maxMilestones`:
1. Should it be mandatory property?
2. What would be the default value if it's mandatory?

**Comment by maxceem** [comment]
3. > Update PATCH endpoint of milestone to respect the maxMilestones value from the timeline

If I understand correctly we should add the same check to the POST endpoint of the milestone.

**Comment by vikasrohit** [comment]
Yes. Agreed. I meant that only. I think we don't need to look for `maxMilestones` for the `PATCH`.

**Comment by vikasrohit** [comment]
> Should it be mandatory property?

I think we don't need to make it required. For `null` values we can just restrict the addition of new milestones because as of today, we don't allow addition/deletion of milestone to any of the product template. And in future, if we need to by pass this check i.e. we need to allow addition of unlimited number of milestones, we can use sentinel value `-1` to denote that state.

**Comment by vikasrohit** [comment]
@maxceem we need to pick this up as well.

**Comment by vikasrohit** [comment]
Moving to backlog as we might not need to restrict the number of milestones per timeline as new designs.
